### PR TITLE
gh-98831: rewrite RAISE_VARARGS in the instruction definition DSL

### PR DESCRIPTION
**Pull request by iritkatriel**
* Issue: gh-98831

### CODE REVIEW

File: Python/bytecodes.c [hunk 1]
```diff
@@ -505,27 +505,26 @@ dummy_func(
...
                 /* fall through */
             case 1:
-                exc = POP(); /* exc */
+                exc = args[0];
                 /* fall through */
             case 0:
                 if (do_raise(tstate, exc, cause)) {
-                    goto exception_unwind;
+                    ERROR_IF(true, exception_unwind);
                 }
```
Comment (gvanrossum) [inline Python/bytecodes.c]: Couldn't this whole block be be this?
```cc
ERROR_IF(do_raise(tstate, exc, cause), exception_unwind);
```

File: Python/generated_cases.c.h [hunk 2]
```diff
@@ -689,25 +689,28 @@
...
                 break;
             default:
                 _PyErr_SetString(tstate, PyExc_SystemError,
                                  "bad RAISE_VARARGS oparg");
                 break;
             }
-            goto error;
+            if (true) { STACK_SHRINK(oparg); goto error; }
+            STACK_SHRINK(oparg);
+            DISPATCH();
```
Comment (iritkatriel) [inline Python/generated_cases.c.h]: The generator emits some unreachable code here. If we had an unconditional ERROR macro we could avoid that.

Comment (gvanrossum) [reply Python/generated_cases.c.h]: The generator could special-case `true` as the argument. There are already other things it checks for to skip generating such dead code (mostly `return` or `goto` or macros containing those), we could add this.

Do you feel it's important enough to do it now? The C compiler or LTO will most likely delete that code anyway, so it doesn't really cost much.

**Review by gvanrossum — COMMENTED** [review]
(Seems I forgot to send one comment.)

**Review by gvanrossum — APPROVED** [review]

File: Tools/cases_generator/test_generator.py [hunk 3]
```diff
@@ -333,20 +333,21 @@ def test_super_instruction():
...
+        op(OP2, (extra/2, arg2, interim -- res)) {
+            res = op2(arg2, interim);
         }
         macro(OP) = OP1 + cache/2 + OP2;
     """
     output = """
         TARGET(OP1) {
-            PyObject *arg = PEEK(1);
+            PyObject *arg1 = PEEK(1);
+            PyObject *interim;
```
Comment (gvanrossum) [inline Tools/cases_generator/test_generator.py]: Note that if you run the tests it'll fail here. I accidentally hit return in the middle of this line and didn't re-run these tests before pushing. As these tests aren't run automatically yet this didn't trigger any CI failures. :-(

I'll fix it in my next change to this file, but I figured I'd warn you.

**Review by gvanrossum — APPROVED** [review]

File: Tools/cases_generator/test_generator.py (hunk 3 above)
Comment (iritkatriel) [reply Tools/cases_generator/test_generator.py]: ok. I also made a mess of my branch just now. Will fix in a bit.

Comment (iritkatriel) [reply Tools/cases_generator/test_generator.py]: We also need to update the makefile. Currently when I change the opcode_metadata.h file I need to touch compile.c. I had a look but wasn't sure what to change.

**Pull request merged**